lib/jwt.rb

```diff
@@ -122,24 +122,27 @@ def decode(jwt, key = nil, verify = true, custom_options = {}, &keyfinder)
     merged_options = DEFAULT_OPTIONS.merge(custom_options)
     decoder = Decode.new jwt, key, verify, merged_options, &keyfinder
     header, payload, signature, signing_input = decoder.decode_segments
-    decode_verify_signature(key, header, signature, signing_input, merged_options, &keyfinder) if verify
+    decode_verify_signature(key, header, payload, signature, signing_input, merged_options, &keyfinder) if verify
     decoder.verify
 
     raise(JWT::DecodeError, 'Not enough or too many segments') unless header && payload
 
     [payload, header]
   end
 
-  def decode_verify_signature(key, header, signature, signing_input, options, &keyfinder)
-    algo, key = signature_algorithm_and_key(header, key, &keyfinder)
+  def decode_verify_signature(key, header, payload, signature, signing_input, options, &keyfinder)
+    algo, key = signature_algorithm_and_key(header, payload, key, &keyfinder)
     if options[:algorithm] && algo != options[:algorithm]
       raise JWT::IncorrectAlgorithm, 'Expected a different algorithm'
     end
     verify_signature(algo, key, signing_input, signature)
   end
 
-  def signature_algorithm_and_key(header, key, &keyfinder)
-    key = yield(header) if keyfinder
+  def signature_algorithm_and_key(header, payload, key, &keyfinder)
+    if keyfinder
+      key = (keyfinder.arity == 2) ? keyfinder.call(header, payload) : keyfinder.call(header)
```

Comment:
Please change your code to use `yield` instead of keyfinder.call (`yield(header, playload)` and `yield(header)`) and use a full `if else` condition statement instead of your ternary conditions.

```ruby
key = if keyfinder.arity == 2
          yield(header, payload)
        else
          yield(header)
```